src/compiler/parser.ts

```diff
@@ -556,98 +558,111 @@ module ts {
         function setDecoratorContext(val: boolean) {
             setContextFlag(val, ParserContextFlags.Decorator);
         }
-
-        function doOutsideOfContext<T>(flags: ParserContextFlags, func: () => T): T {
-            let currentContextFlags = contextFlags & flags;
-            if (currentContextFlags) {
-                setContextFlag(false, currentContextFlags);
+        
+        function setAwaitContext(val: boolean) {
+            setContextFlag(val, ParserContextFlags.Await);
+        }
+        
+        function setAsyncParameterContext(val: boolean) {
+            setContextFlag(val, ParserContextFlags.AsyncParameter);
+        }
+        
+        function doOutsideOfContext<T>(context: ParserContextFlags, func: () => T): T {
+            let setContextFlags = context & contextFlags;
+            if (setContextFlags) {
+                setContextFlag(false, setContextFlags);
                 let result = func();
-                setContextFlag(true, currentContextFlags);
+                setContextFlag(true, setContextFlags);

```

Comment:
is it intentional that you setContextFlag to setContextFlags both before and after calling 'func'?  That seems fishy to me.  If it's intentional definitely document in the code why this is correct.